Add order source attribution to the checkout block

tomalec opened the pull request:
This is addition to https://github.com/woocommerce/woocommerce/pull/39701

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

- [x] Move JS changes...
- [x] Move PHP changes from blocks repo PR
	https://github.com/woocommerce/woocommerce-blocks/pull/11506 
- [x] Remove jQuery dependency from the checkout frontend
	
	
![image](https://github.com/woocommerce/woocommerce/assets/17435/977d01a7-c856-4743-bbf1-0011d1b87265)



### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout the branch and build
2. Add items to the cart
3. go the the block-based checkout
4. Open network tab in the dev tools
5. "Place order"
5. Check the presence and contents of `extensions["woocommerce/order-source-attribution"]` in the payload of `/wp-json/wc/store/v1/checkout` request
6. Check the order metadata is saved correctly.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

### Additional notes
1. I added the JS file too the <code>/client/<strong>legacy</strong></code> even though. It's a piece of vanilla JS targeting the **modern** blocks. But I didn't find a better place.

tomalec committed:
Move changes from blocks repo PR

https://github.com/woocommerce/woocommerce-blocks/pull/11506
Co-authored-by: Jeremy Pry <jeremy.pry@gmail.com>

tomalec committed:
Move changes from blocks repo PR

https://github.com/woocommerce/woocommerce-blocks/pull/11506
Co-authored-by: Jeremy Pry <jeremy.pry@gmail.com>

tomalec committed:
Fix OSA.js dependency ids and checkout extension prefix

layoutd commented:
@tomalec, can you confirm that the payload matches the schema? `SourceAttributionBlocksController::get_schema_callback()` returns this schema (which matches the hidden `<input>` field names added to the classic checkout):
<details>

```json
{
   "wc_order_source_attribution_type":{
      "description":"Source attribution field: type",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_url":{
      "description":"Source attribution field: url",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_campaign":{
      "description":"Source attribution field: utm_campaign",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_source":{
      "description":"Source attribution field: utm_source",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_medium":{
      "description":"Source attribution field: utm_medium",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_content":{
      "description":"Source attribution field: utm_content",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_id":{
      "description":"Source attribution field: utm_id",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_utm_term":{
      "description":"Source attribution field: utm_term",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_session_entry":{
      "description":"Source attribution field: session_entry",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_session_start_time":{
      "description":"Source attribution field: session_start_time",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_session_pages":{
      "description":"Source attribution field: session_pages",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_session_count":{
      "description":"Source attribution field: session_count",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   },
   "wc_order_source_attribution_user_agent":{
      "description":"Source attribution field: user_agent",
      "type":[
         "string",
         "null"
      ],
      "context":[ ],
      "arg_options":{
         "validate_callback":{ },
         "sanitize_callback":{ }
      }
   }
}
```
</details>

But the payload for `extensions[woocommerce/order-source-attribution]` looks like this:
<details>

```json
{
    "current": {
        "typ": "utm",
        "src": "google",
        "mdm": "cpc",
        "cmp": "y_test_sourcebuster_js",
        "cnt": "(none)",
        "trm": "sbjstrm",
        "id": "(none)"
    },
    "current_add": {
        "fd": "2023-11-17 14:11:41",
        "ep": "https://woo.store/wp-content/public/?utm_source=google&utm_medium=cpc&utm_campaign=y_test_sourcebuster_js&utm_term=sbjstrm",
        "rf": "https://woo.store/wp-content/public/"
    },
    "first": {
        "typ": "typein",
        "src": "(direct)",
        "mdm": "(none)",
        "cmp": "(none)",
        "cnt": "(none)",
        "id": "(none)",
        "trm": "(none)",
        "mtke": "(none)"
    },
    "first_add": {
        "fd": "2023-11-11 02:39:10",
        "ep": "https://woo.store/shop/",
        "rf": "https://woo.store/wp-admin/"
    },
    "session": {
        "pgs": "8",
        "cpg": "https://woo.store/checkout/"
    },
    "udata": {
        "vst": "8",
        "uip": "(none)",
        "uag": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36"
    },
    "promo": {
        "code": "694538"
    }
}
```

tomalec committed:
Unify OSA script handles

Stick to `wc-` convention instead of `woocommerce-`.
Do not add a superflous `-js` suffix`

tomalec committed:
Prefix and flatten OSA properties for the Blocks checkout

tomalec committed:
Save OSA proerties on Blocks checkout

using the same method we use for the classic checkout.

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

tomalec commented:
Thanks @layoutd for your help. I flattened the structure (sharing the code with the classic checkout)

tomalec committed:
Fix translation context and PHPCS

layoutd commented on plugins/woocommerce/src/Internal/Orders/SourceAttributionBlocksController.php:
```diff
@@ -0,0 +1,245 @@
+		);
+	}
+
+	/**
+	 * Check if WCCom_Cookie_Terms is available.
+	 *
+	 * @return bool
+	 */
+	protected function is_wccom_cookie_terms_available() {
+		return class_exists( WCCom_Cookie_Terms::class );
+	}
```
We can probably remove this method (https://github.com/woocommerce/woocommerce/pull/39701#issuecomment-1820274517)
```suggestion
```

tomalec committed:
Remove not needed `is_wccom_cookie_terms_available`

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

tomalec committed:
Remove jQuery dependency from the checkout frontend

tomalec committed:
Merge branch 'feature/order_source_attribution-blocks' of github.com:woocommerce/woocommerce into feature/order_source_attribution-blocks

tomalec committed:
Fix PHPCS

tomalec commented on plugins/woocommerce/src/Internal/Orders/SourceAttributionBlocksController.php:
```diff
@@ -0,0 +1,237 @@
+	private function register_assets() {
+		// Once Blocks are moved to the monorepo, hopefully, we can use the following code.
+
+		/*
+		$this->asset_api->register_script(
+			'wc-blocks-order-source-attribution',
+			'build/wc-blocks-order-source-attribution.js',
+			[ 'wc-order-source-attribution' ]
+		);
+		*/
```
It seems our linter check blocks even warnings. Should we remove it completely, or can we comment out in a way the checks will pass? I think it could be valuable to use blocks API once it's available.

layoutd replied:
What is the purpose of adding the code snippet in a comment? Just to have it in mind for future use? If so, I would just add a textual comment about migrating to the Asset API once blocks is merged, I think.

tomalec replied:
Yes. To have is just in place to use once blocks is migrated.

tomalec replied:
After a second thought, I see that using block's AssetsApi does not give much benefit here.
I removed all the comments in https://github.com/woocommerce/woocommerce/pull/41592/commits/2932640bfdac4f092579aa769d481df6932348c2

opr commented on plugins/woocommerce/src/Internal/Orders/SourceAttributionBlocksController.php:
```diff
@@ -0,0 +1,237 @@
+				'schema_callback' => $this->get_schema_callback(),
+			)
+		);
+		// The following does not work.
+		// See https://github.com/woocommerce/woocommerce-blocks/blob/d646b42e98dc69fe06356f54b53fd6cff132fe98/docs/third-party-developers/extensibility/rest-api/extend-rest-api-update-cart.md#basic-usage .
+
+		/*
+		$this->extend_schema->register_update_callback(
+			[
+				'namespace' => 'woocommerce/order-source-attribution',
+				'callback'  => function( $data ) {
+					// Save the data to the order here.
+				},
+			]
+		);
+		*/
```
I'd like to understand what's not working about this.

This API is designed to update the WC_Cart - not to modify the order after "Place order" has been pressed. The hook used below is the correct way for this.

Was there some incorrect expectations based on the docs, or code Rubik shared with you that made you try this?

My main concern is giving devs the best DX possible, so any feedback on that point would be great!

tomalec replied:
Thank you for the clarifications! :) 

What made me try this, was that this function was present in https://github.com/woocommerce/woocommerce-blocks/pull/11506 I took over

From the name itself, I read `register_endpoint_data` is to register an endpoint in the API to receive the data, with the schema etc., and `register_update_callback` is to register the callback to update when the data is recieved. 
The name itself does not give a clue on restriction to the cart only. 

Then I searched the code and docs for anything related to `register_endpoint_data` and `register_update_callback`, and found this:
https://github.com/woocommerce/woocommerce-blocks/blob/d646b42e98dc69fe06356f54b53fd6cff132fe98/docs/third-party-developers/extensibility/rest-api/extend-rest-api-update-cart.md#basic-usage which states:
> `ExtendSchema` offers the ability for extensions to register **callback functions to be executed** when signalled to do so **by** the client-side Cart **or Checkout**.

So I thought: " The function name does not seem to be cart-specific; the cart is what will eventually be checked out. Therefore updating the cart should update what will be checked out," - so it made sense to me to use it.
I was a bit unsure, as it is not taking the `'endpoint'` parameter, but I thought maybe it is not needed, and the callback is chosen based on the `namespace` matching.

To double-check, `$this->extend_schema->register_endpoint_data(` is the right way to extend what would be expected on the checkout request?

tomalec replied:
I removed the comment in https://github.com/woocommerce/woocommerce/pull/41592/commits/77889265bccbca59cb752e8f01d86750761df2fa

tomalec committed:
Remove unrelevant code comment

Addresses https://github.com/woocommerce/woocommerce/pull/41592#discussion_r1402411581

tomalec committed:
Simplify `woocommerce_order_save_attribution_source_data` logic

tomalec committed:
Remove code comments related to AssetsAPI

Addresses https://github.com/woocommerce/woocommerce/pull/41592#discussion_r1402021606

tomalec commented on plugins/woocommerce/src/Internal/Traits/SourceAttributionMeta.php:
```diff
@@ -230,19 +230,39 @@ private function get_hpos_order_object( $post_or_order ) {
 	 * @return array
 	 */
-	private function get_source_values( array $raw_values = array() ): array {
+	private function get_unprefixed_fields( array $raw_values = array() ): array {
```
I was not sure whether it's better to keep it here or move to `SourceAttributionController` as it's only used there. On the other hand, all the other `*prefixed_` related methods are here.

tomalec committed:
Blanket Order Source Attribution → Order Attribution change

Merge branch 'feature/order_source_attribution' into feature/order_source_attribution-blocks

tomalec committed:
Blanket Order Source Attribution → Order Attribution for blocks

Propagate chenges from https://github.com/woocommerce/woocommerce/pull/39701/commits/c9c2a05ca5cab352777108332ecc871b3f87fcfb to the Block-related classes

tomalec committed:
Do the classic checkout OSA bindings only on the classic page

Call `previousInitCheckout` only if there is one.

tomalec committed:
fix PHPCS

layoutd commented on plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php:
```diff
@@ -0,0 +1,209 @@
+	 *
+	 * @param ExtendSchema                $extend_schema                 ExtendSchema instance.
+	 * @param FeaturesController          $features_controller           Features controller.
+	 * @param OrderAttributionController  $order_attribution_controller Instance of the order attribution controller.
```
PHPCS fix
```suggestion
	 * @param ExtendSchema               $extend_schema                 ExtendSchema instance.
	 * @param FeaturesController         $features_controller           Features controller.
	 * @param OrderAttributionController $order_attribution_controller Instance of the order attribution controller.
```

layoutd commented on plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php:
```diff
@@ -0,0 +1,209 @@
+
+use Automattic\Jetpack\Constants;
+use Automattic\WooCommerce\Internal\Features\FeaturesController;
+use Automattic\WooCommerce\Internal\Orders\OrderAttributionController;
```
```suggestion
```
Import unnecessary

layoutd reviewed (commented):
Trying to get it to build to fully test, adding some PHPCS comments meanwhile

tomalec committed:
Fix PHPCS

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

tomalec committed:
Remove unnecesary import

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

layoutd commented on plugins/woocommerce/client/legacy/js/frontend/order-attribution.js:
```diff
@@ -31,34 +57,27 @@
+		 * Add source values to the classic checkout.
 		 */
-		$( document.body ).on( 'init_checkout', () => { setFields(); } );
+		if ( $( '.woocommerce form.checkout' ) !== null ) {
```
This doesn't work for classic checkout, there's no form element that matches this selector. This seems to work:
```suggestion
		if ( $( 'form.woocommerce-checkout' ) !== null ) {
```

layoutd replied:
Sorry, it works with `/shortcode → [woocommerce_checkout]` but not with `/Classic Checkout`:
<img width="844" alt="image" src="https://github.com/woocommerce/woocommerce/assets/228780/e27f296d-22b9-45cf-8541-ff5de39f6247">

The alternative seems to work with both

layoutd reviewed (changes requested):
Looks good except for one edge case with the form selector that I came across 👏🏻

tomalec committed:
Update classic checkout selector

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

layoutd commented on plugins/woocommerce/src/Internal/Traits/OrderAttributionMeta.php:
```diff
@@ -230,19 +230,39 @@ private function get_hpos_order_object( $post_or_order ) {
 		// Look through each field in POST data.
 		foreach ( $this->fields as $field ) {
-			$value = sanitize_text_field( wp_unslash( $raw_values[ $this->get_prefixed_field( $field ) ] ?? '(none)' ) );
+			$values[ $field ] = $raw_values[ $this->get_prefixed_field( $field ) ];
```
As you mentioned in discussion, this line needs the `??` null coalescing operator that got moved into `get_source_values` because `$_POST` may not include all these values

tomalec replied:
Thanks, fixed in https://github.com/woocommerce/woocommerce/commit/cda715786ad23b83712ddd126c5d85903fbc7332

tomalec committed:
Move `??` operator from `get_source_values` back to `get_unprefixed_fields`

 because `$_POST` may not include all these values.

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>

layoutd reviewed (approved):
LGTM! We should document what you've mentioned previously about making the mapping of the fields more dynamic (or not) for a future PR.
ETA → Already done: https://github.com/woocommerce/woocommerce/pull/39701#issuecomment-1825792668.
